nixos/modules/services/networking/headscale.nix

```diff
@@ -0,0 +1,486 @@
+            ".*" = "default-namespace";
+          };
+        };
+
+      };
+
+      tls = {
+        letsencrypt = {
+          hostname = mkOption {
+            type = types.nullOr types.str;
+            default = "";
+            description = ''
+              Domain name to request a TLS certificate for.
+            '';
+          };
+          challengeType = mkOption {
+            type = types.enum [ "TLS_ALPN-01" "HTTP-01" ];
+            default = "HTTP-01";
+            description = ''
+              Type of ACME challenge to use, currently supported types:
+              HTTP-01 or TLS_ALPN-01
+            '';
+          };
+          httpListen = mkOption {
+            type = types.nullOr types.str;
+            default = ":http";
+            description = ''
+              When HTTP-01 challenge is chosen, letsencrypt must set up a
+              verification endpoint, and it will be listning on:
+              :http = port 80
+            '';
+          };
+        };
+
+        certFile = mkOption {
+          type = types.nullOr types.path;
+          default = "";
+          description = ''
+            Path to already created certificate.
+          '';
+        };
+        keyFile = mkOption {
+          type = types.nullOr types.path;
+          default = "";
+          description = ''
+            Path to key for already created certificate.
+          '';
+        };
+      };
+
+      aclPolicyFile = mkOption {
+        type = types.nullOr types.path;
+        default = "";
+        description = ''
+          Path to a file containg ACL policies
+        '';
+      };
+
+    };
+
+  };
+  config = mkIf cfg.enable {
+    # Setup the headscale configuration in a known path in /etc to
+    # allow both the Server and the Client use it to find the socket
+    # for communication.
+    environment.etc."headscale/config.yaml".source = configFile;
+
+    users.groups = optionalAttrs (cfg.group == "headscale") {
+      headscale = { };
+    };
+
+    users.users = optionalAttrs (cfg.user == "headscale") {
+      headscale = {
+        description = "headscale user";
+        createHome = true;
+        home = cfg.dataDir;
+        group = cfg.group;
+        useDefaultShell = true;
+        isNormalUser = true;
+      };
+    };
+
+    systemd.services.headscale = {
+      description = "headscale coordination server for Tailscale";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = ''
+        ${optionalString (cfg.database.passwordFile != null) ''
+        HEADSCALE_DB_PASS="$(head -n1 ${cfg.database.passwordFile})"
+        ''}
+
+        HEADSCALE_OIDC_CLIENT_SECRET="$(head -n1 ${cfg.openIdConnect.clientSecretFile})"
+        ${cfg.package}/bin/headscale serve
+      '';
+
+      environment = {
+        # Make sure the systemd configuration reloads if the
+        # configuration has changed
+        HEADSCALE_CONFIG_SHA512 = builtins.hashFile "sha512" configFile;
```

Comment:
`configFile` is a store path, so it already contains a hash. if there's a way to specify an alternate config file in the start script it wold be better to use that mechanism.

Reply:
You're looking for `restartTriggers` - that is the convention for what you are doing here.

Reply:
yes, that would work just as well. (we just think it's nicer to not have services use implicit paths in `/etc` if at all possible)

Reply:
Very happy to take guidance regarding symlinking the file to `/etc`. I tried to clarify why its done in a comment and I have not seen an obvious way to work around it to make the `headscale` CLI "find" it when a user is running it.

Will look into the restartTriggers.